cql3/restrictions/statement_restrictions.cc

```diff
@@ -1057,6 +1061,359 @@ ::shared_ptr<expression> make_conjunction(::shared_ptr<expression> a, ::shared_p
+    } else {
+        throw std::logic_error("empty tuple on LHS of =");
+    }
+}
+
+/// True iff lhs is limited by rhs in the manner prescribed by op.
+bool limits(const bytes& lhs, const operator_type& op, const bytes& rhs, const abstract_type& type) {
+    if (!op.is_compare()) {
+        throw std::logic_error("limits() called on non-compare op");
+    }
+    const auto cmp = type.as_tri_comparator()(lhs, rhs);
+    if (cmp < 0) {
+        return op == operator_type::LT || op == operator_type::LTE || op == operator_type::NEQ;
+    } else if (cmp > 0) {
+        return op == operator_type::GT || op == operator_type::GTE || op == operator_type::NEQ;
+    } else {
+        return op == operator_type::LTE || op == operator_type::GTE || op == operator_type::EQ;
+    }
+}
+
+/// True iff the value of opr.lhs is limited by opr.rhs in the manner prescribed by opr.op.
+bool limits(const binary_operator& opr, const selection& selection, row_data& cells,
+            const query_options& options) {
+    if (!opr.op.is_slice()) { // For EQ or NEQ, use equals().
+        throw std::logic_error("limits() called on non-slice op");
+    }
+    const auto& columns = std::get<0>(opr.lhs);
+    if (columns.size() > 1) {
+        auto multi = dynamic_pointer_cast<multi_item_terminal>(opr.rhs);
+        if (!multi) {
+            throw std::logic_error("RHS for multi-column is not a tuple");
+        }
+        const auto& rhs = multi->get_elements();
+        if (rhs.size() != columns.size()) {
+            throw std::logic_error("LHS and RHS size mismatch");
+        }
+        for (size_t i = 0; i < rhs.size(); ++i) {
+            const auto cmp = comparator(columns[i])->as_tri_comparator()(
+                    *get_value(columns[i], selection, cells, options),
+                    *rhs[i]);
+            // If the components aren't equal, then we just learned the LHS/RHS order.
+            if (cmp < 0) {
+                if (opr.op == operator_type::LT || opr.op == operator_type::LTE) {
+                    return true;
+                } else if (opr.op == operator_type::GT || opr.op == operator_type::GTE) {
+                    return false;
+                } else {
+                    throw std::logic_error("Unknown slice operator");
+                }
+            } else if (cmp > 0) {
+                if (opr.op == operator_type::LT || opr.op == operator_type::LTE) {
+                    return false;
+                } else if (opr.op == operator_type::GT || opr.op == operator_type::GTE) {
+                    return true;
+                } else {
+                    throw std::logic_error("Unknown slice operator");
+                }
+            }
+            // Otherwise, we don't know the LHS/RHS order, so check the next component.
+        }
+        // Getting here means LHS == RHS.
+        return opr.op == operator_type::LTE || opr.op == operator_type::GTE;
```

Comment:
A possible future optimization is to compare the upper and lower limits simultaneously, reducing duplicate get_value() calls.

Reply:
Exactly, that's what I'm aiming for.  I'm just not there yet; replacing old code gradually, from inside out.